### PR TITLE
Query node status and report if not stable

### PR DESCRIPTION
**Pull request by MartinMinkov**
**Description**

Check the specified URL for the node status before deploying. If the node is completely offline or gives a bad response, we return early. If the node is in either `CONNECTING`, `BOOTSTRAP`, or `CATCHUP`, we also report an error and return early.

**Tested By**
Manual testing.

### CODE REVIEW

File: src/lib/deploy.js [hunk 1]
```diff
@@ -157,6 +157,32 @@ async function deploy({ network, yes }) {
...
+  const graphQLEndpoint = config?.networks[network]?.url ?? DEFAULT_GRAPHQL;
+  const { data: nodeStatus } = await sendGraphQL(
+    graphQLEndpoint,
+    getStatusQuery()
+  );
+
+  if (!nodeStatus || nodeStatus.syncStatus === 'OFFLINE') {
+    log(
+      red(
+        `  Failed to make a stable connection to the specified URL. Please try again.`
```
[the remark this answers is not in the document]
Comment (ymekuria) [reply src/lib/deploy.js:169]: One suggestion would be to add the actual url that had a failed connection in the error message with a template string.

File: src/lib/deploy.js [hunk 2]
```diff
@@ -157,6 +157,32 @@ async function deploy({ network, yes }) {
...
+    );
+    return;
+  } else if (
+    nodeStatus.syncStatus === 'CONNECTING' ||
+    nodeStatus.syncStatus === 'BOOTSTRAP' ||
+    nodeStatus.syncStatus === 'CATCHUP'
+  ) {
+    log(
+      red(
+        `  The specified URL node status is not ready for zkApp deployments. Please try again.`
```
[the remark this answers is not in the document]
Comment (ymekuria) [reply src/lib/deploy.js:180]: That's a good question. I think we can probably lump them into one failed sync status to reduce cognitive overhead for the user, unless you can think of a case where a user would care about the different sync statuses.

**Review by ymekuria — APPROVED** [review]
LGTM. I just left a couple minor comments replying to your questions.

File: src/lib/deploy.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ymekuria) [reply src/lib/deploy.js:169]: Maybe something like 'Failed to make a stable connection to the node specified at ${graphQLEndpoint}. Please try again.`

[the remark this answers is not in the document]
Comment (jasongitmail) [reply src/lib/deploy.js:169]: Suggestion: `Transaction relayer node is offline. Please try again or use a different "url" for this network alias in your config.json`

File: src/lib/deploy.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasongitmail) [reply src/lib/deploy.js:180]: Suggestion: `Transaction relayer node is not in a synced, ready state. Its status is "${nodeStatus.syncStatus}". Please try again when the node is synced or use a different "url" for this network alias in your config.json`

I'm not  100% on "synced, ready" but maybe it's good. wdyt @MartinMinkov ?

File: src/lib/deploy.js [hunk 3]
```diff
@@ -500,6 +525,13 @@ function getAccountQuery(publicKey) {
   }`;
 }
 
+function getStatusQuery() {
+  return `
+  query {
```
Comment (jasongitmail) [inline src/lib/deploy.js:530]: Not sure it's worth separating this out. Would it make sense to put this in Line 163?

**Review by jasongitmail — APPROVED** [review]

File: src/lib/deploy.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasongitmail) [reply src/lib/deploy.js]: I like the version you went with @MartinMinkov 👍

**Comment by jasongitmail** [comment]
LGTM!

**Pull request merged**